### PR TITLE
nrf_rpc: Initialize transports independently of each other

### PR DESCRIPTION
**Pull request by maje-emb**
This adds support to initialize transports independently. 
If one transport fails to bond, the others are attempted to initialize.

### CODE REVIEW

**Review by grochu — CHANGES_REQUESTED** [review]

File: nrf_rpc/include/nrf_rpc.h [hunk 1]
```diff
@@ -108,6 +108,7 @@ struct nrf_rpc_group_data {
 	uint8_t src_group_id;
 	uint8_t dst_group_id;
 	struct nrf_rpc_os_event decode_done_event;
+	bool transport_used;
```
[the remark this answers is not in the document]
Comment (maje-emb) [reply nrf_rpc/include/nrf_rpc.h]: Good suggestion. I have changed the name to `transport_initialized`.

File: nrf_rpc/nrf_rpc.c [hunk 2]
```diff
@@ -337,25 +352,29 @@ static int transport_init(nrf_rpc_tr_receive_handler_t receive_cb)
...
 			}
 		}
 
+		group->data->transport_used = true;
 		err = group_init_send(group);
 		if (err) {
 			NRF_RPC_ERR("Failed to send group init packet for group id: %d strid: %s",
 				    data->src_group_id, group->strid);
-			return err;
+			continue;
```
[the remark this answers is not in the document]
Comment (maje-emb) [reply nrf_rpc/nrf_rpc.c:370]: To be able to initialize the group, the transport must first be initialized. Therefore, the set variable is before the `if` block.
After handshake, the group will be initialized, `dst_group_id` will be assigned.
In this PR, the `NRF_RPC_GROUP_STATUS` macro was added, which is used to check the group status.

File: nrf_rpc/include/nrf_rpc_cbor.h [hunk 3]
```diff
@@ -124,6 +124,14 @@ struct nrf_rpc_cbor_ctx {
 		_nrf_rpc_cbor_prepare((struct nrf_rpc_cbor_ctx *)(&(_ctx)), (_len) + 1); \
 	} while (0)
 
+/** @brief Checks memory for a packet.
+ * 
+ * Mecro checks whether memory for a packet has been allocated.
+ * 
+ * @param[in] _ctx Context allocated by @ref NRF_RPC_CBOR_ALLOC.
+ */
+#define NRF_RPC_CBOR_IS_ALLOC(_ctx) ((_ctx).out_packet != NULL)
```
Comment (kapi-no) [inline nrf_rpc/include/nrf_rpc_cbor.h:133]: What is the purpose of this change? It seems unrelated to the main change. If it's a standalone API improvement, you should move it to a separate commit.

**Review by kapi-no — COMMENTED** [review]
LGTM, just one nitpick

File: nrf_rpc/include/nrf_rpc_cbor.h (hunk 3 above)
Comment (maje-emb) [reply nrf_rpc/include/nrf_rpc_cbor.h:133]: Buffor allocation is done by RPC transport function. If the transport is not properly initialized the buffer allocation macro - `NRF_RPC_CBOR_ALLOC`, does not return error. Therefore, it is a helper macro. 

In addition, there is also macro `NRF_RPC_GROUP_STATUS` to check whether the group/transport was initialized properly. If not then the user should not use such a group.

File: nrf_rpc/nrf_rpc.c [hunk 4]
```diff
@@ -105,17 +105,32 @@ static void free_rx_buf(const struct nrf_rpc_group *group, const uint8_t *packet
...
 static int send(const struct nrf_rpc_tr *transport, const uint8_t *data, size_t length)
 {
+	struct nrf_rpc_group *group;
+
 	if (!transport->api->send) {
 		NRF_RPC_ASSERT(false);
 		return -NRF_EIO;
 	}
 
+	group = NRF_RPC_CONTAINER_OF(transport, struct nrf_rpc_group, transport);
```
[the remark this answers is not in the document]
Comment (maje-emb) [reply nrf_rpc/nrf_rpc.c]: Changed. `send` function takes a `group` as parameter instead of `transport`.

**Review by grochu — APPROVED** [review]

File: nrf_rpc/nrf_rpc.c [hunk 5]
```diff
@@ -337,25 +352,29 @@ static int transport_init(nrf_rpc_tr_receive_handler_t receive_cb)
...
 		err = group_init_send(group);
 		if (err) {
 			NRF_RPC_ERR("Failed to send group init packet for group id: %d strid: %s",
 				    data->src_group_id, group->strid);
-			return err;
+			continue;
 		}
 	}
 
 	err = nrf_rpc_os_event_wait(&groups_init_event, CONFIG_NRF_RPC_GROUP_INIT_WAIT_TIME);
```
[the remark this answers is not in the document]
Comment (maje-emb) [reply nrf_rpc/nrf_rpc.c:370]: I'm not sure if this is needed. The `transport_init` function has to wait for the initialization procedure of all groups in both cases. This function does not return information  about status of a particular group.
Checking if a group is initialized correctly can be done by using the `NRF_RPC_GROUP_STATUS`.

**Comment by maje-emb** [comment]
I added an entry to `CHANGELOG` file.

File: nrf_rpc/include/nrf_rpc_cbor.h [hunk 6]
```diff
@@ -124,6 +124,14 @@ struct nrf_rpc_cbor_ctx {
 		_nrf_rpc_cbor_prepare((struct nrf_rpc_cbor_ctx *)(&(_ctx)), (_len) + 1); \
 	} while (0)
 
+/** @brief Checks memory for a packet.
+ * 
+ * Mecro checks whether memory for a packet has been allocated.
```
Comment (peknis) [inline nrf_rpc/include/nrf_rpc_cbor.h:129]:
```suggestion
 * Macro checks whether memory for a packet has been allocated.
```

File: nrf_rpc/CHANGELOG.rst [hunk 7]
```diff
@@ -19,6 +19,11 @@ Added
 
 * Added the :kconfig:option:`CONFIG_NRF_RPC_ZCBOR_BACKUPS` Kconfig option for setting up the number of zcbor backups.
 
+Changes
+=======
+
+* The initialization of transport is independent of each other.
```
Comment (peknis) [inline nrf_rpc/CHANGELOG.rst:25]:
```suggestion
* Different transport mechanisms are initialized independent of each other.
```
Hoping not to change the meaning here.

**Review by peknis — CHANGES_REQUESTED** [review]

File: nrf_rpc/CHANGELOG.rst (hunk 7 above)
Comment (maje-emb) [reply nrf_rpc/CHANGELOG.rst]: This looks okay to me. Thanks.

**Review by anhmolt — APPROVED** [review]

File: nrf_rpc/CHANGELOG.rst [hunk 8]
```diff
@@ -19,6 +19,11 @@ Added
 
 * Added the :kconfig:option:`CONFIG_NRF_RPC_ZCBOR_BACKUPS` Kconfig option for setting up the number of zcbor backups.
 
+Changes
+=======
+
+* Different transport mechanisms are initialized independent of each other.
```
Comment (grochu) [inline nrf_rpc/CHANGELOG.rst:25]: Could you make this entry a bit more precise: e.g.:
```suggestion
* Added a possibility for transport in each group to initialize independently of other groups. This enables the RPC to work for groups which are ready earlier rather than waiting for all groups to initialize their transports.
```
or something similar.

Comment (maje-emb) [reply nrf_rpc/CHANGELOG.rst]: Thanks for the suggestion.
I have changed the second part of the entry. Because the `nrf_rpc_init` function waits for all groups to be initialized. After that it is possible to use RPC.

File: nrf_rpc/CHANGELOG.rst [hunk 9]
```diff
@@ -19,6 +19,12 @@ Added
 
 * Added the :kconfig:option:`CONFIG_NRF_RPC_ZCBOR_BACKUPS` Kconfig option for setting up the number of zcbor backups.
 
+Changes
+=======
+
+* Added a possibility for transport in each group to initialize independently of other groups.
+* This enables the RPC to work for those groups which are ready, even though initialization of others may have failed.
```
Comment (peknis) [inline nrf_rpc/CHANGELOG.rst:26]:
```suggestion
* Enabled independent initialization of transport in each group.
  This makes RPC available for the groups that are ready, even though initialization of others may have failed.
```
Not separate bullets, because this is very much cause and effect.

**Review by peknis — CHANGES_REQUESTED** [review]

File: nrf_rpc/CHANGELOG.rst (hunk 9 above)
Comment (maje-emb) [reply nrf_rpc/CHANGELOG.rst]: Right.

**Review by peknis — APPROVED** [review]

File: nrf_rpc/include/nrf_rpc_cbor.h [hunk 10]
```diff
@@ -124,6 +124,14 @@ struct nrf_rpc_cbor_ctx {
 		_nrf_rpc_cbor_prepare((struct nrf_rpc_cbor_ctx *)(&(_ctx)), (_len) + 1); \
 	} while (0)
 
+/** @brief Checks memory for a packet.
+ * 
+ * Macro checks whether memory for a packet has been allocated.
+ * 
+ * @param[in] _ctx Context allocated by @ref NRF_RPC_CBOR_ALLOC.
+ */
+#define NRF_RPC_CBOR_IS_ALLOC(_ctx) ((_ctx).out_packet != NULL)
```
Comment (KAGA164) [inline nrf_rpc/include/nrf_rpc_cbor.h:133]: Could it be a function?

Comment (maje-emb) [reply nrf_rpc/include/nrf_rpc_cbor.h]: Yes, it can be a function.

**Review by KAGA164 — APPROVED** [review]
